packages/kbn-ui-theme/src/theme.ts

```diff
@@ -17,19 +17,26 @@ export type Theme = typeof v8Light;
 
 // in the Kibana app we can rely on this global being defined, but in
 // some cases (like jest) the global is undefined
+/** @deprecated theme can be dynamic now, access is discouraged */
 export const tag: string = globals.__kbnThemeTag__ || 'v8light';
+/** @deprecated theme can be dynamic now, access is discouraged */
 export const version = 8;
+/** @deprecated theme can be dynamic now, access is discouraged */
 export const darkMode = tag.endsWith('dark');
 
 export const euiLightVars: Theme = v8Light;
 export const euiDarkVars: Theme = v8Dark;
 
+let isDarkMode = darkMode;
+export const _setDarkMode = (mode: boolean) => {
+  isDarkMode = mode;
+};
+
 /**
  * EUI Theme vars that automatically adjust to light/dark theme
  */
-export let euiThemeVars: Theme;
-if (darkMode) {
-  euiThemeVars = euiDarkVars;
-} else {
-  euiThemeVars = euiLightVars;
-}
+export const euiThemeVars: Theme = new Proxy(isDarkMode ? euiDarkVars : euiLightVars, {
+  get(accessedTarget, accessedKey, ...rest) {
+    return Reflect.get(isDarkMode ? euiDarkVars : euiLightVars, accessedKey, ...rest);
+  },
+});
```

Comment:
This is for point `5. euiThemeVars` from the issue's description

Core's browser-side `theme` service now leverages the new `_setDarkMode` setter when darkMode is `system` to dynamically swap the target of the `euiThemeVars` proxy. This is done soon enough in the page load's chain (before plugins are loaded), so we're sure that there's no references from plugin codes toward properties of the proxy before core gets to call `_setDarkMode`.